Fix memory allocation on STM32L4 devices

c1728p9 opened the pull request:
### Description

Depending on initial size allocated on STM32L4 devices with TWO_RAM_REGIONS  set a crash may occur. This is because there is a mismatch between the size newlib is expecting and the size actually returned by _sbrk. This is because the STM32L4 implementation of _sbrk is performing alignment internally.

This patch fixes this problem by removing the code in __wrap__sbrk which performs the alignment.

### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

c1728p9 committed:
Fix memory allocation on STM32L4 devices

Depending on initial size allocated on STM32L4 devices with
TWO_RAM_REGIONS  set a crash may occur. This is because there is a
mismatch between the size newlib is expecting and the size actually
returned by _sbrk. This is because the STM32L4 implementation of _sbrk
is performing alignment internally.

This patch fixes this problem by removing the code in __wrap__sbrk
which performs the alignment.

adustm commented on targets/TARGET_STM/TARGET_STM32L4/l4_retarget.c:
```diff
@@ -50,8 +48,8 @@ extern uint32_t __mbed_krbs_start;
+    uint32_t heap_ind_old = heap_ind;
+    uint32_t heap_ind_new = heap_ind_old + incr;
 
     if (heap_ind_new > &__mbed_krbs_start) {
```
Hi, could we take the opportunity of removing a compilation warning here ?
```
[Warning] l4_retarget.c@54,22: comparison between pointer and integer
[DEBUG] Return: 0
[DEBUG] Output: .\targets\TARGET_STM\TARGET_STM32L4\l4_retarget.c: In function '__wrap__sbrk':
[DEBUG] Output: .\targets\TARGET_STM\TARGET_STM32L4\l4_retarget.c:54:22: warning: comparison between pointer and integer
[DEBUG] Output:      if (heap_ind_new > &__mbed_krbs_start) {
[DEBUG] Output:                       ^
```

With
```
-    if (heap_ind_new > &__mbed_krbs_start) {
+    if (heap_ind_new > (uint32_t)&__mbed_krbs_start) {
```

0xc0170 replied:
Thanks for catching this warning 💯

adustm commented:
cc @betzw (I think you might appreciate this ;)  )

Hi @c1728p9 
Many thanks for this modification. I tested it ok for #7867 
Kind regards

betzw commented:
@adustm :smile:

MateuszMaz commented:
Hi @c1728p9 , as you requested I tested it. It works for l475 and l476. I also checked all the tests that mbed-os and ci test shield provides - everything is ok about this fix, no regression anywhere

c1728p9 committed:
Fix comparison warning on STM32L4 devices

Cast the pointer used in l4_retarget to uint32_t before comparing it
to fix the warning:
"comparison between pointer and integer"

c1728p9 commented:
Fixed warnings as well. @cmonr can you make sure this gets into the next RC?

cmonr commented:
Marked for RC2 for now, but this doesn't look like it was discovered during OOB.
@adbridge Thoughts?

adbridge commented:
@cmonr We should only be bringing in essential fixes for RC2. As this has not been found to be an issue so far we should probably push it to 5.10.1

jeromecoutant commented:
> this has not been found to be an issue so far

This is maybe because there is no STM32L4 devices with TWO_RAM_REGIONS in the CI :-)

adustm commented:
Hello @adbridge 
Every STM32L476 / STM32L475 have issues on the current release about memory mapping.

Please refer to  https://github.com/ARMmbed/mbed-os/issues/7867
I'm wondering wether this issue is related or not
https://os.mbed.com/questions/82162/Cannot-get-SD-card-working-with-NUCLEO-L/

Kind regards

adbridge commented:
Moving this back to RC2 so that we can get the fix into 5.10 as it looks like this has a bigger impact than originally thought.

adbridge commented:
/morph build

cmonr commented:
Test failure does not appear to be related to PR.

/morph test

cmonr commented:
Test failure does not appear to be related to PR.
/morph test

cmonr commented:
/morph export-build
